vnext/ReactUWP/Views/KeyboardEventHandler.cpp

```diff
@@ -0,0 +1,402 @@
+{
+  DispatchEventToJs("topKeyDown", args);
+}
+
+void PreviewKeyboardEventHandlerOnRoot::OnPreKeyUp(winrt::IInspectable const& sender, winrt::KeyRoutedEventArgs const&args)
+{
+  DispatchEventToJs("topKeyUp", args);
+}
+
+HandledKeyboardEventHandler::HandledKeyboardEventHandler()
+{
+}
+
+void HandledKeyboardEventHandler::UpdateHandledKeyboardEvents(string propertyName, folly::dynamic const& value)
+{
+  if (propertyName == "keyDownEvents") {
+    m_handledKeyDownKeyboardEvents = KeyboardHelper::FromJS(value);
+  }
+  else if (propertyName == "keyUpEvents")
+  {
+    m_handledKeyUpKeyboardEvents = KeyboardHelper::FromJS(value);
+  }
+}
+
+void HandledKeyboardEventHandler::hook(XamlView xamlView)
+{
+  unhook();
+
+  EnsureKeyboardEventHandler();
+  m_previewKeyboardEventHandler->hook(xamlView);
+  m_keyboardEventHandler->hook(xamlView);
+}
+
+void HandledKeyboardEventHandler::unhook()
+{
+  if (m_previewKeyboardEventHandler)
+    m_previewKeyboardEventHandler->unhook();
+  if (m_keyboardEventHandler)
+    m_keyboardEventHandler->unhook();
+}
+
+void HandledKeyboardEventHandler::EnsureKeyboardEventHandler()
+{
+  if (!m_previewKeyboardEventHandler)
+  {
+    m_previewKeyboardEventHandler = make_unique<PreviewKeyboardEventHandler>(
+      std::bind(&HandledKeyboardEventHandler::KeyboardEventHandledHandler, this, KeyboardEventPhase::PreviewKeyDown, _1, _2),
+      std::bind(&HandledKeyboardEventHandler::KeyboardEventHandledHandler, this, KeyboardEventPhase::PreviewKeyUp, _1, _2)
+      );
+  }
+
+  if (!m_keyboardEventHandler)
+  {
+    m_keyboardEventHandler = make_unique<KeyboardEventHandler>(
+      std::bind(&HandledKeyboardEventHandler::KeyboardEventHandledHandler, this, KeyboardEventPhase::KeyDown, _1, _2),
+      std::bind(&HandledKeyboardEventHandler::KeyboardEventHandledHandler, this, KeyboardEventPhase::KeyUp, _1, _2)
+      );
+  }
+}
+
+void HandledKeyboardEventHandler::KeyboardEventHandledHandler(KeyboardEventPhase phase, winrt::IInspectable const& sender, winrt::KeyRoutedEventArgs const& args)
+{
+  HandledEventPhase currentEventPhase = (phase == KeyboardEventPhase::PreviewKeyUp || phase == KeyboardEventPhase::PreviewKeyDown)
+    ? HandledEventPhase::CAPTURING : HandledEventPhase::BUBBLING;
+
+  auto ev = KeyboardHelper::CreateKeyboardEvent(currentEventPhase, args);
+
+  bool shouldMarkHandled = false;
+  if (phase == KeyboardEventPhase::PreviewKeyDown || phase == KeyboardEventPhase::KeyDown)
+    shouldMarkHandled = ShouldMarkKeyboardHandled(m_handledKeyDownKeyboardEvents, ev);
+  else
+    shouldMarkHandled = ShouldMarkKeyboardHandled(m_handledKeyUpKeyboardEvents, ev);
+
+  if (shouldMarkHandled)
+    args.Handled(true);
+}
+
+bool HandledKeyboardEventHandler::ShouldMarkKeyboardHandled(std::vector<KeyboardEvent> const& handledEvents, KeyboardEvent currentEvent)
+{
+  for (auto event : handledEvents)
+  {
+    if (event.key == currentEvent.key &&
+      (!event.altKey || (event.altKey && currentEvent.altKey)) &&
```

Comment:
>!event.altKey [](start = 7, length = 13)

So, if the event spec doesn't say to use the alt key we will accept an event with alt or not?

Reply:
I send out email to discuss it.

---
In reply to: [295542258](https://github.com/microsoft/react-native-windows/pull/2623#discussion_r295542258) [](ancestors = 295542258)

Reply:
This is already tracked by email. I would submit another PR if I finally need to make change on it.
No change here.

Reply:
As discussed in the email thread, let's change this to an explicit model:  the modifier key states must match exactly for us to mark the event as handled.

---
In reply to: [295592610](https://github.com/microsoft/react-native-windows/pull/2623#discussion_r295592610) [](ancestors = 295592610)

Reply:
Done